### PR TITLE
Add max_extrinsics_limit for package transactions

### PR DESCRIPTION
**Pull request by gguoss**
Thank you for your Pull Request!

Before you submitting, please check that:

- [ ] You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points reviewers should know?
  - Is there something left for follow-up PRs?
- [ ] You labeled the PR with appropriate labels if you have permissions to do so.
- [ ] You mentioned a related issue if this PR related to it, e.g. `Fixes #228` or `Related #1337`.
- [ ] You asked any particular reviewers to review. If you aren't sure, start with GH suggestions.
- [x] Your PR adheres [the style guide](https://github.com/paritytech/polkadot/wiki/Style-Guide)
  - In particular, mind the maximal line length.
  - There is no commented code checked in unless necessary.
  - Any panickers have a proof or removed.
- [ ] You bumped the runtime version if there are breaking changes in the **runtime**.
- [ ] You updated any rustdocs which may have changed

After you've read this notice feel free to remove it.
Thank you!

✄ -----------------------------------------------------------------------------

### CODE REVIEW

**Comment by gguoss** [comment]
> What's the use case of this? We already have a limit that takes the size of extrinsics into account, what's the point of limiting by count?
> 
> Note even if we are willing to accept that, it should rather be implemented fully in runtime (`executive`) and not leak into `basic_authorship`

Case: When there are too many transactions in the trading pool, it will affect the nodes can not pack too many transactions (such as 10,000 tx, of course, the total size is less than 4M (you call the default control node total data size)). Causes a large number of nodes to have no such performance processing, resulting in too long long. This affects overall network performance.

So add max_extrinsics_limit for tx number. The number of packaged transactions can be adjusted through the parliament according to the network.

**Comment by xlc** [comment]
A better way is modify executive module. You can have it check extrinsic index and return `Err(internal::ApplyError::FullBlock)` if it is more than a number to indicate block full

https://github.com/paritytech/substrate/blob/19e533b3b01a0dab2d1238baaf223b348e4c24d4/srml/executive/src/lib.rs#L282

**Comment by gguoss** [comment]
> A better way is modify executive module. You can have it check extrinsic index and return `Err(internal::ApplyError::FullBlock)` if it is more than a number to indicate block full
> 
> https://github.com/paritytech/substrate/blob/19e533b3b01a0dab2d1238baaf223b348e4c24d4/srml/executive/src/lib.rs#L282

Accept

File: srml/executive/src/lib.rs [hunk 1]
```diff
@@ -251,6 +253,13 @@ where
 			return Err(internal::ApplyError::FullBlock);
 		}
 
+		let extrinsic_count =
+			storage::unhashed::get(well_known_keys::MAX_EXTRINSICS_LIMIT).unwrap_or(internal::MAX_EXTRINSICS_LIMIT);
+		// Check the extrinsics count of the block if that extrinsic is applied.
+		if <system::Module<System>>::extrinsic_count()  > extrinsic_count {
```
Comment (gavofyork) [inline srml/executive/src/lib.rs]:
```suggestion
		if <system::Module<System>>::extrinsic_count() > extrinsic_count {
```

**Comment by gavofyork** [comment]
i'm not keen on the additional key lookup - it's an extra weird "known key" and it's extra trie i/o each block/transaction. i would instead prefer to stick with the const value in executive alone and if you want to change it, then recompile and do a runtime-upgrade.

**Comment by bkchr** [comment]
@gguoss any update?

**Comment by gguoss** [comment]
> i'm not keen on the additional key lookup - it's an extra weird "known key" and it's extra trie i/o each block/transaction. i would instead prefer to stick with the const value in executive alone and if you want to change it, then recompile and do a runtime-upgrade.

Accept

**Comment by bkchr** [comment]
I thought again about this. We should probably make the `Executive` configurable and make a trait, something like `BlockFullness` or whatever. A type implementing this trait can be passed to the `Executive` by using the configuration trait.
This makes it possible for every runtime to have a custom "when is my block full".

**Comment by tomusdrw** [comment]
I suppose `Weight` (#2796) covers this particular use case where count limit would be necessary. I like the idea of having this customisable in runtime, but it feels a bit like overengineering currently.

**Comment by bkchr** [comment]
Yeah, because of the `weight` stuff I had this idea :)

**Comment by Kianenigma** [comment]
Weight stuff does cover this but not in a fully configurable manner. 

One can decide what the weight representation is + define what is the max. 

Rest of the logic + linking it to how much fee a particular tx with weight `x` has is currently hardcoded into executive. 

> I thought again about this. We should probably make the Executive configurable and make a trait, something like BlockFullness or whatever. A type implementing this trait can be passed to the Executive by using the configuration trait.
This makes it possible for every runtime to have a custom "when is my block full".

This is great and I actually did a bit drafting on it, but I wonder if it is possible to enable this input object to be _state-full_? Looks like an overdue but something along the line of _give it the system trait as some generic_?

**Comment by bkchr** [comment]
I will close this pr, if I have done this to early, sorry. Just reopen it :)

**Pull request closed without merging**